### PR TITLE
git: Apply sub-repo permissions to git.ArchiveReader

### PR DESCRIPTION
**Pull request by sashaostrikov**
Since we don't have a valid use case for an end-user to request an entire archive of a repo, sub-repo permissions check returns an error if the context indicates this is a user request.

Closes https://github.com/sourcegraph/sourcegraph/issues/27332

### CODE REVIEW

File: internal/vcs/git/archive.go [hunk 1]
```diff
@@ -20,7 +24,18 @@ const (
...
+func ArchiveReader(
+	ctx context.Context,
+	checker authz.SubRepoPermissionChecker,
+	repo api.RepoName,
+	format ArchiveFormat,
+	commit api.CommitID,
+	relativePath string,
+) (io.ReadCloser, error) {
+	a := actor.FromContext(ctx)
+	if checker != nil && checker.Enabled() && a.IsAuthenticated() {
```
Comment (mollylogue) [inline internal/vcs/git/archive.go:36]: You could use the function I just added for this now if you want. It's fine as is too.

**Review by mollylogue — APPROVED** [review]
Looks good to me! I like the approach of just simplifying things by not allowing access to this function if sub-repo permissions are enabled.

**Review by ryanslade — APPROVED** [review]

File: internal/vcs/git/archive.go (hunk 1 above)
Comment (sashaostrikov) [reply internal/vcs/git/archive.go]: absolutely necessary, thanks! I don't want to create any code duplication 😄

**Comment by sashaostrikov** [comment]
The only change I made is to use Molly's function. All checks are green, therefore I guess I'm good to merge it now

**Pull request merged**